main.tf

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/******************************************
+	Scheduled Function Definition
+ *****************************************/
+
+resource "google_cloud_scheduler_job" "job" {
+  provider = "google-beta"
+  name     = "${var.job_name}"
+  description = "${var.job_description}"
+  schedule = "${var.job_schedule}"
+
+  pubsub_target = {
+    topic_name = "projects/${var.project_id}/topics/${module.pubsub_topic.topic}"
+    data = "${var.message_data}"
+  }
+}
+
+/******************************************
+	PubSub Topic Definition
+ *****************************************/
+
+module "pubsub_topic" {
+  source  = "github.com/terraform-google-modules/terraform-google-pubsub"
+  topic   = "${var.topic_name}"
+  project_id = "${var.project_id}"
+}
+
+/******************************************
+	Cloud Function Resource Definitions
+ *****************************************/
+
+resource "google_cloudfunctions_function" "main" {
+  provider = "google-beta"
+  name                  = "${var.name}"
+  source_archive_bucket = "${google_storage_bucket.main.name}"
+  source_archive_object = "${google_storage_bucket_object.main.name}"
+  description           = "${var.function_description}"
+  available_memory_mb   = "${var.function_available_memory_mb}"
+  timeout               = "${var.function_timeout_s}"
+  entry_point           = "${var.function_entry_point}"
+
+  event_trigger {
+    event_type = "google.pubsub.topic.publish"
+    resource   = "${module.pubsub_topic.topic}"
+
+    failure_policy {
+      retry = "${var.function_event_trigger_failure_policy_retry}"
+    }
+  }
+
+  labels                = "${var.function_labels}"
+  runtime               = "${var.function_runtime}"
+  environment_variables = "${var.function_environment_variables}"
+  project               = "${var.project_id}"
+  region                = "${var.region}"
+}
+
+data "archive_file" "main" {
+  type        = "zip"
+  output_path = "${pathexpand("${var.function_source_directory}.zip")}"
+  source_dir  = "${pathexpand("${var.function_source_directory}")}"
+}
+
+resource "random_string" "random_suffix" {
+  length  = 4
+  upper   = "false"
+  special = "false"
+}
+
+resource "google_storage_bucket" "main" {
+  provider = "google-beta"
+  name          = "${var.name}"
```

Comment:
This is user specified at the moment though. Adding a random suffix I think would be a bit of unexpected behaviour for a user that specified the exact name they wanted. I'd vote that if we the random suffix, we should either not let the user set it, or make it so the default is random and the user could override. Thoughts?

Reply:
I like what you're pitching @ogreface . Assigning a default value of the bucket name with randomization should probably be the default but an explicit override should be specifiable.

Buckets are in the global name space for all projects. My opinion (emphasis on opinion) is to typically build buckets with:

`${var.bucket_prefix}-${random_id.suffix.result}-${data.google_project.project.number}`

Better still, `substr` that result with the `min` of 63 (max length) and that interpolated string to ensure you haven't tipped over the edge.

* The bucket prefix is important to make sense of what a bucket is for.
* The final bits of randomization mean within a given project, there's little chance for collision.
* The project number is a good way to ensure uniqueness across customers.

Even with 4 characters of randomization should do fine. Regardless, I think we're partially in the realm of opinion here and should really come to a standard within CFT.

Reply:
This is now updated. I separated name into function_name, and bucket_name. bucket_name will be used as is if passed in, otherwise it becomes a concatenation of <project id>-scheduled-function-XXXX where XXXX are random chars.